app/events/applications/application_deleted_event.rb

```diff
@@ -9,7 +9,8 @@ def self.create(application)
       metadata: {
         provider_id: application.provider_account_id || application.tenant_id,
         zync: {
-          service_id: application.service_id
+          service_id: application.service_id,
+          proxy_id: application.service&.proxy&.id # TODO: It could not exist anymore!
```

Comment:
I do not think we would need `proxy_id`

Reply:
Yup 😄

Reply:
We would need to check in zync if it deletes correctly the application if the proxy was already deleted.

Reply:
@hallelujah That makes sense 💪 but how can we do that?

Reply:
![image](https://user-images.githubusercontent.com/11318903/50594751-13200d00-0e9e-11e9-8710-2e281a1b161b.png)

Reply:
Ok, I think there is some info that needs to be added here and so I will 😄 I definitely need help to check this though 😞 Because I am having problems... no matter what I do, I don't get the client deleted in keycloak 😕 event without removing the Proxy first 😞 

### Steps to check it works
1. Install and configure everything (Zync, Keycloak, and the integration with Porta/System). I will explain below how to do this 😄 
2. Delete the proxy manually with `delete`. Going to Porta's `rails console` and doing `Account.find(2).default_service.proxy.detele`
3. Destroy the app. Through Porta's `rails console`, I removed them all, just in case, first the developer's ones with `Account.find(2).provided_cinstances.find_each { |app| app.destroy! }` and later its own cinstance as a buyer `Account.find(2).bought_cinstances.find_each { |app| app.destroy! }`
4. See if it works. This means checking in keycloak if the client has been removed.

### Why
#### Summary of Zync works for Apps
Zync receives events from Porta, in this case a delete event, then it fetches Porta to see if the app is still there. If it isn't, then it deletes in keycloak.
Same general way of working for create and for update.
#### Relationship with its Service and Proxy
** I WILL IMPROVE THIS EXPLANATION **
> Well there are some cases
> for example zync does not have the service created yet
> so this worker create the parents for this object to be valid
> imagine:
> 
> 1. I push a zync event for application but service and proxy are not created yet in zync
> 2. then I get a UnprocessableEntityError
> 3. given that I just push 2 events for the dependencies of the application: Service and proxy
> So service and proxy will be created inside zync database in the next iteration
> my question is can we work without proxy in zync to delete an application?
> Not to create obviously
> 
> Marta Noya [11:09 AM]
> To delete, ok, got you :+1:
> Apparently not, but I am not sure yet, let me do a test :slightly_smiling_face:


### How to install and configure everything
```markdown
# Local integration between Porta/System with Keycloak through Zync

## Install Porta
Follow [the install guide](https://github.com/3scale/porta/blob/master/INSTALL.md).

## Install Zync
Download [Zync](https://github.com/3scale/zync) and run `bundle install`.

Install PostgreSQL with `brew install postgresql`
Start its service with `brew services start postgresql`
Make sure that it is running. See the services running with `brew services list` and `postgresql` should be in that list with status `started` and looking green. If you see an orange `orange` started, that means that it is trying to start but it is having errors, which you can see with more detail in the logs with `tail -f /usr/local/var/log/postgres.log`.

Create the database with `rails db:create`.

## Install Keycloak
You can have a keycloak running locally with `docker run --name keycloak -d -p 8080:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=p jboss/keycloak`
So that docker container with keycloak will be accessible from your browser in `http://localhost:8080`.

After that, you will need to follow the instructions to [configure Red Hat Single Sign-On](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/openid-connect).

## Run Zync
From the folder where you have Zync installed, run: `ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec rails server -p 5000` so you will get zync running in the port 5000.

## Run Porta
If your porta is already running, stop it, and if you have `spring`, also do `spring stop`.
You will need these configurations:
  - In `config/rolling_updates.yml`, make sure you have `apicast_v2: true` and `apicast_oidc: true`.
  - In `config/sandbox_proxy.yml`, make sure you have `apicast_registry_url: <%= ENV.fetch('APICAST_REGISTRY_URL', '<Insert a URL that contains policies in JSON. It can be am empty JSON for this test in particular>') %>`

Run porta with `UNICORN_WORKERS=8 ZYNC_ENDPOINT=http://localhost:5000 ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec rails server -b 0.0.0.0`

## Configure 3scale to use Keycloak
Follow the documentation to [configure 3scale](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/openid-connect).
```

Reply:
We checked it and we think we also reproduced [the bug "THREESCALE-1480 - Client ID not registered in RH-SSO"](https://issues.jboss.org/browse/THREESCALE-1480).
Anyway, we could bypass this problem as [mentioned in this comment of the issue](https://issues.jboss.org/browse/THREESCALE-1480?focusedCommentId=13680818&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13680818).

We have discovered that the `proxy_id` needs to be there as well in order to work, so what we are going to do for this is every time that we delete a proxy, we create a `ProxyDeletedEvent`, and we fetch the id from there if the proxy is already deleted when the `ApplicationDeletedEvent` is created.

Reply:
I guess that I've got to fix the System/Porta side so far (although I still have to check the efficiency of it!!), but in Zync, I don't get it to work even without removing the Proxy first 😭 
These are the logs:
```
Started PUT "/notification" for ::1 at 2019-01-14 15:58:55 +0100
Processing by NotificationsController#update as JSON
  Parameters: {"type"=>"Application", "id"=>5, "parent_event_id"=>"d15b5ec1-bd8a-4120-a65c-6988db01afcc", "parent_event_type"=>"Applications::ApplicationDeletedEvent", "tenant_id"=>2, "service_id"=>2, "proxy_id"=>2, "notification"=>{"id"=>5, "tenant_id"=>2}}
  Tenant Load (0.2ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
  ↳ app/controllers/notifications_controller.rb:20
   (0.2ms)  BEGIN
  ↳ app/services/incoming_notification_service.rb:15
   (0.1ms)  SAVEPOINT active_record_1
  ↳ app/models/application_record.rb:9
  Application Load (0.4ms)  SELECT  "applications".* FROM "applications" WHERE "applications"."id" = $1 AND "applications"."tenant_id" = $2 AND "applications"."service_id" = $3 AND "applications"."tenant_id" = $4 LIMIT $5  [["id", 5], ["tenant_id", 2], ["service_id", 2], ["tenant_id", 2], ["LIMIT", 1]]
  ↳ app/models/notification.rb:40
  Model Load (0.4ms)  SELECT  "models".* FROM "models" WHERE "models"."record_type" = $1 AND "models"."record_id" = $2 AND "models"."tenant_id" = $3 LIMIT $4  [["record_type", "Application"], ["record_id", 5], ["tenant_id", 2], ["LIMIT", 1]]
  ↳ app/models/model.rb:23
   (0.2ms)  RELEASE SAVEPOINT active_record_1
  ↳ app/models/application_record.rb:9
  Notification Create (0.6ms)  INSERT INTO "notifications" ("model_id", "data", "tenant_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["model_id", 5], ["data", "{\"type\":\"Application\",\"id\":5,\"parent_event_id\":\"d15b5ec1-bd8a-4120-a65c-6988db01afcc\",\"parent_event_type\":\"Applications::ApplicationDeletedEvent\",\"tenant_id\":2,\"service_id\":2,\"proxy_id\":2,\"format\":\"json\",\"controller\":\"notifications\",\"action\":\"update\",\"notification\":{\"id\":5,\"tenant_id\":2}}"], ["tenant_id", 2], ["created_at", "2019-01-14 14:58:55.041253"], ["updated_at", "2019-01-14 14:58:55.041253"]]
  ↳ app/services/incoming_notification_service.rb:18
[ActiveJob] Enqueued UpdateJob (Job ID: a0ffbdf6-fbeb-4044-9f7d-b9243d181be9) to Que(default) with arguments: #<GlobalID:0x00007fcae6da9388 @uri=#<URI::GID gid://zync/Model/5>>
   (0.4ms)  COMMIT
  ↳ app/services/incoming_notification_service.rb:15
Completed 201 Created in 14ms (Views: 0.4ms | ActiveRecord: 2.4ms)


  Model Load (0.3ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
  ↳ /Users/marta/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-5.2.2/lib/active_record/log_subscriber.rb:98
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9] Performing UpdateJob (Job ID: a0ffbdf6-fbeb-4044-9f7d-b9243d181be9) from Que(default) with arguments: #<GlobalID:0x00007fcae5d26088 @uri=#<URI::GID gid://zync/Model/5>>
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]    (0.2ms)  BEGIN
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/models/update_state.rb:9
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   Model Load (0.3ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2 FOR NO KEY UPDATE NOWAIT  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/models/model.rb:28
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   UpdateState Load (0.6ms)  SELECT  "update_states".* FROM "update_states" WHERE "update_states"."model_id" = $1 LIMIT $2 FOR UPDATE  [["model_id", 5], ["LIMIT", 1]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/models/update_state.rb:10
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   Model Load (0.3ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/jobs/update_job.rb:23
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   UpdateState Update (0.4ms)  UPDATE "update_states" SET "started_at" = $1, "updated_at" = $2 WHERE "update_states"."id" = $3  [["started_at", "2019-01-14 14:58:55.699362"], ["updated_at", "2019-01-14 14:58:55.701240"], ["id", 5]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/jobs/update_job.rb:23
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   Application Load (0.3ms)  SELECT  "applications".* FROM "applications" WHERE "applications"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/services/fetch_service.rb:24
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   Tenant Load (0.2ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/services/fetch_service.rb:56
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9] [FetchService] got Entry (Application): {"tenant_id"=>2, "model_id"=>5}
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   Entry Create (0.5ms)  INSERT INTO "entries" ("tenant_id", "model_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["tenant_id", 2], ["model_id", 5], ["created_at", "2019-01-14 14:58:55.856803"], ["updated_at", "2019-01-14 14:58:55.856803"]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/jobs/update_job.rb:27
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   UpdateState Update (0.4ms)  UPDATE "update_states" SET "finished_at" = $1, "updated_at" = $2 WHERE "update_states"."id" = $3  [["finished_at", "2019-01-14 14:58:55.858529"], ["updated_at", "2019-01-14 14:58:55.859017"], ["id", 5]]
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/jobs/update_job.rb:27
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]    (1.4ms)  COMMIT
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9]   ↳ app/models/update_state.rb:9
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9] Enqueued ProcessEntryJob (Job ID: f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3) to Que(default) with arguments: #<GlobalID:0x00007fcae6d0ea68 @uri=#<URI::GID gid://zync/Entry/25>>
[ActiveJob] [UpdateJob] [a0ffbdf6-fbeb-4044-9f7d-b9243d181be9] Performed UpdateJob (Job ID: a0ffbdf6-fbeb-4044-9f7d-b9243d181be9) from Que(default) in 167.91ms
{"lib":"que","hostname":"Martas-MacBook-Pro","pid":17217,"thread":70254716834180,"event":"job_worked","job":{"queue":"","priority":100,"run_at":"2019-01-14 14:58:55 +0000","job_id":53,"job_class":"ActiveJob::QueueAdapters::QueAdapter::JobWrapper","args":[{"job_class":"UpdateJob","job_id":"a0ffbdf6-fbeb-4044-9f7d-b9243d181be9","provider_job_id":null,"queue_name":"default","priority":null,"arguments":[{"_aj_globalid":"gid://zync/Model/5"}],"executions":0,"locale":"en"}],"error_count":0},"elapsed":0.17503}
  Entry Load (0.3ms)  SELECT  "entries".* FROM "entries" WHERE "entries"."id" = $1 LIMIT $2  [["id", 25], ["LIMIT", 1]]
  ↳ /Users/marta/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-5.2.2/lib/active_record/log_subscriber.rb:98
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3] Performing ProcessEntryJob (Job ID: f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3) from Que(default) with arguments: #<GlobalID:0x00007fcae6cdbcf8 @uri=#<URI::GID gid://zync/Entry/25>>
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   Model Load (0.2ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/jobs/process_entry_job.rb:15
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]    (0.1ms)  BEGIN
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/models/application_record.rb:9
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   Application Load (0.4ms)  SELECT  "applications".* FROM "applications" WHERE "applications"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/jobs/process_entry_job.rb:18
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   Tenant Load (0.2ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/models/integration.rb:13
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   Service Load (0.2ms)  SELECT  "services".* FROM "services" WHERE "services"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/models/application.rb:7
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   Model Load (0.4ms)  SELECT  "models".* FROM "models" WHERE "models"."record_type" = $1 AND "models"."record_id" = $2 LIMIT $3  [["record_type", "Service"], ["record_id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/models/application.rb:7
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]    (0.2ms)  COMMIT
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/models/application_record.rb:9
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   Integration Load (0.4ms)  SELECT "integrations".* FROM "integrations" WHERE ("integrations"."tenant_id" = $1 AND "integrations"."model_id" IS NULL OR "integrations"."model_id" = $2)  [["tenant_id", 2], ["model_id", 4]]
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3]   ↳ app/jobs/process_entry_job.rb:27
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3] Enqueued ProcessIntegrationEntryJob (Job ID: 85bf3dcf-96df-42ed-b739-d33ea006c012) to Que(default) with arguments: #<GlobalID:0x00007fcae63230a8 @uri=#<URI::GID gid://zync/Integration::Keycloak/1>>, #<GlobalID:0x00007fcae6321bb8 @uri=#<URI::GID gid://zync/Model/5>>
[ActiveJob] [ProcessEntryJob] [f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3] Performed ProcessEntryJob (Job ID: f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3) from Que(default) in 10.2ms
{"lib":"que","hostname":"Martas-MacBook-Pro","pid":17217,"thread":70254716834180,"event":"job_worked","job":{"queue":"","priority":100,"run_at":"2019-01-14 14:58:55 +0000","job_id":54,"job_class":"ActiveJob::QueueAdapters::QueAdapter::JobWrapper","args":[{"job_class":"ProcessEntryJob","job_id":"f8fa50ad-26be-4ea7-8cd5-6c4a68ab5bd3","provider_job_id":null,"queue_name":"default","priority":null,"arguments":[{"_aj_globalid":"gid://zync/Entry/25"}],"executions":0,"locale":"en"}],"error_count":0},"elapsed":0.01568}
  Integration::Keycloak Load (0.4ms)  SELECT  "integrations".* FROM "integrations" WHERE "integrations"."type" IN ('Integration::Keycloak') AND "integrations"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  ↳ /Users/marta/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-5.2.2/lib/active_record/log_subscriber.rb:98
  Model Load (0.3ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
  ↳ /Users/marta/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-5.2.2/lib/active_record/log_subscriber.rb:98
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012] Performing ProcessIntegrationEntryJob (Job ID: 85bf3dcf-96df-42ed-b739-d33ea006c012) from Que(default) with arguments: #<GlobalID:0x00007fcae6cae8e8 @uri=#<URI::GID gid://zync/Integration::Keycloak/1>>, #<GlobalID:0x00007fcae6caca98 @uri=#<URI::GID gid://zync/Model/5>>
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]    (0.1ms)  BEGIN
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/integration_state.rb:11
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   IntegrationState Load (0.4ms)  SELECT  "integration_states".* FROM "integration_states" WHERE "integration_states"."model_id" = $1 AND "integration_states"."integration_id" = $2 LIMIT $3 FOR UPDATE  [["model_id", 5], ["integration_id", 1], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/integration_state.rb:12
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Entry Load (0.3ms)  SELECT  "entries".* FROM "entries" WHERE "entries"."model_id" = $1 ORDER BY "entries"."id" DESC LIMIT $2  [["model_id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/entry.rb:31
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Application Load (0.2ms)  SELECT  "applications".* FROM "applications" WHERE "applications"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/jobs/process_integration_entry_job.rb:86
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Model Load (0.2ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/jobs/process_integration_entry_job.rb:42
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Integration Load (0.2ms)  SELECT  "integrations".* FROM "integrations" WHERE "integrations"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/jobs/process_integration_entry_job.rb:42
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   IntegrationState Update (0.3ms)  UPDATE "integration_states" SET "started_at" = $1, "entry_id" = $2, "success" = $3, "updated_at" = $4 WHERE "integration_states"."id" = $5  [["started_at", "2019-01-14 14:58:55.892089"], ["entry_id", 25], ["success", nil], ["updated_at", "2019-01-14 14:58:55.895586"], ["id", 3]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/jobs/process_integration_entry_job.rb:42
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Model Load (0.2ms)  SELECT  "models".* FROM "models" WHERE "models"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/entry.rb:8
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Application Load (0.2ms)  SELECT  "applications".* FROM "applications" WHERE "applications"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/entry.rb:8
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Tenant Load (0.2ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/services/integration/keycloak_service.rb:42
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Service Load (0.2ms)  SELECT  "services".* FROM "services" WHERE "services"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/services/integration/keycloak_service.rb:43
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   CACHE Tenant Load (0.0ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/client.rb:11
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   IntegrationState Update (0.4ms)  UPDATE "integration_states" SET "success" = $1, "finished_at" = $2, "updated_at" = $3 WHERE "integration_states"."id" = $4  [["success", true], ["finished_at", "2019-01-14 14:58:55.902257"], ["updated_at", "2019-01-14 14:58:55.902651"], ["id", 3]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/jobs/process_integration_entry_job.rb:38
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   Tenant Load (0.2ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/jobs/process_integration_entry_job.rb:131
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]    (0.4ms)  COMMIT
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012]   ↳ app/models/integration_state.rb:11
[ActiveJob] [ProcessIntegrationEntryJob] [85bf3dcf-96df-42ed-b739-d33ea006c012] Performed ProcessIntegrationEntryJob (Job ID: 85bf3dcf-96df-42ed-b739-d33ea006c012) from Que(default) in 20.57ms
{"lib":"que","hostname":"Martas-MacBook-Pro","pid":17217,"thread":70254716834180,"event":"job_worked","job":{"queue":"","priority":100,"run_at":"2019-01-14 14:58:55 +0000","job_id":55,"job_class":"ActiveJob::QueueAdapters::QueAdapter::JobWrapper","args":[{"job_class":"ProcessIntegrationEntryJob","job_id":"85bf3dcf-96df-42ed-b739-d33ea006c012","provider_job_id":null,"queue_name":"default","priority":null,"arguments":[{"_aj_globalid":"gid://zync/Integration::Keycloak/1"},{"_aj_globalid":"gid://zync/Model/5"}],"executions":0,"locale":"en"}],"error_count":0},"elapsed":0.02796}
```

Reply:
Good news! I've done a workaround for Zync to work locally, then I tested it... It works even with proxy_id being nil 😄 It still removes the client from keycloak 💪